### PR TITLE
Default column order for setcolorder based on keys

### LINKED ISSUE #2895: Default colorder with setcolorder based off keys and indices
I'm proposing to have `setcolorder(x)` set the columns in the order specified by the keys, then remaining index columns, then the rest of the columns. I have a fork with the change already. Not sure if it would be welcome though. It makes for a nice mindless but sensible ordering to the data.

```R
DT = data.table(a = 1:3, b = 2:4, c = 3:5)
setkey(DT, c)
setindex(DT, b)
all.equal(names(setcolorder(DT)),  c("c", "b", "a"))
```

Forked branch available here https://github.com/Rdatatable/data.table/compare/master...jsams:colorder_keys

Happy to make a PR if this is something that would be welcome (and sorry for the tracker spam the past few hours). Tried setting up the tests and such, but really, I have no idea what's going on with the testing framework here to be honest.

### PR DESCRIPTION
**Pull request by jsams**
With this PR, `setcolorder` can now, when called without further arguments, set the columns in the order specified by the keys, then remaining index columns, then the rest of the columns.

Closes #2895

### CODE REVIEW

**Comment by MichaelChirico** [comment]
delete `.swp` files. I would rebase so that the `.swp` doesn't enter history, even.

I would also add `*.swp` to `.gitignore`, in fact; somewhat surprised it hasn't made it in there yet.

File: NEWS.md [hunk 1]
```diff
@@ -7,6 +7,10 @@
 
 1. Empty RHS of := is no longer an error when the i clause returns no rows to assign to anyway, [#2829](https://github.com/Rdatatable/data.table/issues/2829). Thanks to @cguill95 for reporting and to @MarkusBonsch for fixing.
 
+#### NEW FEATURES
+
+1. `setcolorder(DT)` without further arguments now defaults to ordering by keys, remaining index columns, then preserving the order of the remaining columns. Thanks to @jsams for the PR.
```
Comment (MichaelChirico) [inline NEWS.md]: Please add the issue # & link (see directly above for reference format)

File: R/data.table.R [hunk 2]
```diff
@@ -2554,8 +2554,17 @@ setnames <- function(x,old,new) {
   invisible(x)
 }
 
-setcolorder <- function(x, neworder)
+setcolorder <- function(x, neworder=NULL)
 {
+  if (is.null(neworder)) {
+    keys = key(x)
+	if (is.null(indices(x))) {
```
Comment (MichaelChirico) [inline R/data.table.R]: annoying, i know, but i think we settled on 2-space indentation for this project 😬

File: man/setcolorder.Rd [hunk 3]
```diff
@@ -13,7 +13,7 @@ setcolorder(x, neworder)
 }
 \arguments{
   \item{x}{ A \code{data.table}. }
-  \item{neworder}{ Character vector of the new column name ordering. May also be column numbers. If \code{length(neworder) < length(x)}, the specified columns are moved in order to the "front" of \code{x}. }
+  \item{neworder}{ Character vector of the new column name ordering. May also be column numbers. If \code{length(neworder) < length(x)}, the specified columns are moved in order to the "front" of \code{x}. If \code{neworder} is \code{NULL}, columns are ordered first by the key columns, remaining index columns, with remaining columns in their current order to the right. }
```
Comment (MichaelChirico) [inline man/setcolorder.Rd]: grammar here -- I'm not fully clear what the new sentence is conveying

File: man/setcolorder.Rd [hunk 4]
```diff
@@ -13,7 +13,7 @@ setcolorder(x, neworder)
 }
```
Comment (MichaelChirico) [inline man/setcolorder.Rd]: `\usage{}` section will have to be updated as well

File: R/data.table.R [hunk 5]
```diff
@@ -2554,8 +2554,17 @@ setnames <- function(x,old,new) {
   invisible(x)
 }
 
-setcolorder <- function(x, neworder)
+setcolorder <- function(x, neworder=NULL)
```
Comment (jangorecki) [inline R/data.table.R]: any reason why we change default to NULL? we can use `missing(neworder)` and argument as is. Or, what I would prefer, is to use `neworder=key(x)` but then we are not using columns that have indices. Then behavior is very clear from manual or `args(setcolorder)`.

File: R/data.table.R (hunk 2 above)
Comment (jangorecki) [reply R/data.table.R]: annoying is when you are losing half of screen because of 4 space indentation :)

File: man/setcolorder.Rd (hunk 4 above)
Comment (jangorecki) [reply man/setcolorder.Rd]: must, otherwise R CMD check warns, best to run check before submitting PR as issues like this can be easily spotted.

**Comment by MichaelChirico** [comment]
I agree with Jan. @jsams are you particularly married to including the indices? What's your thinking on using indices, so we're clear?

**Comment by jsams** [comment]
*facepalm* cannot believe I added the swp files in. And then the two space tabs... Guess that's why your not supposed to commit just before bed ;p

Happy to wait to here from Matt on his thoughts. Was a little surprised things moved as quickly as they did actually.

re: also using indices, I agree that the keys are the key portion of this FR. I didn't have any particular use case in mind for the indices. I was trying to think about if I was going to impose a default ordering on the tables, what would that look like. It seemed to me that index columns might have slightly higher priority than non-index columns. I also thought that maybe alphabetizing the non-key columns might make sense.

Again, my data isn't usually particularly wide (a couple of dozen columns at most 99% of the time); so I don't have great intuition what's a convenient way to organize the rest of those columns. I just know I like my keys first. :D

**Comment by mattdowle** [comment]
I guess a lot of people prefer key columns first, including me. So I'm generally in agreement with this PR.  It would be nice to visually separate the key columns from the others with a vertical bar `|` in `print.data.table()` and for that, the key columns need to be first.
Does `setcolorder(DT)` (with no other arguments) do anything currently?
In addition to (or instead of) `setcolorder(setkey(DT, cols))`, how about adding an argument to `setkey()` like `setkey(DT, ..., moveFirst=TRUE)`, or a new function `setkeyfirst()`, `setKeyFirst()`, `setKEY()` or `setkey2()` or something like that?

File: R/data.table.R [hunk 6]
```diff
@@ -2554,8 +2554,17 @@ setnames <- function(x,old,new) {
...
+	if (is.null(indices(x))) {
+		indcs = c()
+	} else {
+		indcs = unique(setdiff(do.call(c, indices(x, vectors=T)), keys))
+	}
+    neworder = c(keys, indcs, setdiff(names(x), c(keys, indcs)))
+  }
   if (any(duplicated(neworder))) stop("neworder contains duplicates")
   # if (!is.data.table(x)) stop("x is not a data.table")
   if (length(neworder) != length(x)) {
```
Comment (MichaelChirico) [inline R/data.table.R]: @mattdowle `setcolorder(x)` in current master will error here. so no old behavior broken

**Comment by MichaelChirico** [comment]
> It would be nice to visually separate the key columns from the others with a vertical bar | in print.data.table() and for that, the key columns need to be first.

@mattdowle this _was_ part of #1523 but was dropped. I don't remember well why, but a quick glance at the discussions suggests it was because distinguishing the order of keys is difficult in general.

This PR could potentially help on that front. Don't want to clutter this up too much w discussion about that issue though.

**Comment by jsams** [comment]
> In addition to (or instead of) setcolorder(setkey(DT, cols)), how about adding an argument to setkey() like setkey(DT, ..., moveFirst=TRUE), or a new function setkeyfirst(), setKeyFirst(), setKEY() or setkey2() or something like that?

I think I would (weakly) prefer `setkey(DT, ..., order=T/F)` or `reorder` or similar, just to keep the name more consistent with `setcolorder`.

The way you've written that suggests you might would want that to default to `TRUE`. For my own code, I would welcome that. But that might interact badly with `rbind(list)` with `use.names=F`.

**Comment by mattdowle** [comment]
Yes I think default for `setkey()` to move key columns first and print nicely with `|` separator, would be good. Not the name `order` for that argument to `setkey()` though. Rather some name more clearly to do with the key columns moving to be first, such as `keyFirst=getOption("data.table.keyFirst", FALSE|TRUE)`.  I doubt breakage would be huge, but there would clearly be some breakage; e.g. reports and any code using column numbers (which isn't recommended and could be detected and warned about).  Could manage the breakage in the usual way with an option and slow migration, warnings etc, but probably a bit much to bite off at the current time.  Not sure why you mentioned `rblindlist(list)` with `use.names=F` specifically.

**Comment by MichaelChirico** [comment]
LGTM, can you copy-paste the error you're getting from `R CMD check`?

**Comment by jsams** [comment]
I couldn't find a useful message. This was all there was:

> * checking PDF version of manual ... WARNING
LaTeX errors when creating PDF version.
This typically indicates Rd problems.

File: man/setcolorder.Rd [hunk 7]
```diff
@@ -9,11 +9,11 @@
...
 
 \usage{
-setcolorder(x, neworder)
+setcolorder(x, neworder=key(x))
 }
 \arguments{
   \item{x}{ A \code{data.table}. }
-  \item{neworder}{ Character vector of the new column name ordering. May also be column numbers. If \code{length(neworder) < length(x)}, the specified columns are moved in order to the "front" of \code{x}. }
+  \item{neworder}{ Character vector of the new column name ordering. May also be column numbers. If \code{length(neworder) < length(x)}, the specified columns are moved in order to the "front" of \code{x}. By default, setcolorder without a specified \code{neworder} moves the key columns in order to the "front" of \code{x}. }
```
Comment (MichaelChirico) [inline man/setcolorder.Rd]: this isn't the warning, but just noticed: wrap `setcolorder` in `\code{}`

**Pull request merged**